api/middleware/logger.go

```diff
@@ -0,0 +1,137 @@
+package middleware
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	opentracing "github.com/opentracing/opentracing-go"
+	log "github.com/sirupsen/logrus"
+	jaeger "github.com/uber/jaeger-client-go"
+	macaron "gopkg.in/macaron.v1"
+)
+
+var (
+	logHeaders = false
+)
+
+type LoggingResponseWriter struct {
+	macaron.ResponseWriter
+	errBody []byte // the body in case it is an error
+}
+
+func (rw *LoggingResponseWriter) Write(b []byte) (int, error) {
+	if rw.ResponseWriter.Status() >= 400 {
+		rw.errBody = make([]byte, len(b))
+		copy(rw.errBody, b)
+	}
+	return rw.ResponseWriter.Write(b)
+}
+
+func Logger() macaron.Handler {
+	return func(ctx *Context) {
+		start := time.Now()
+		ctx.Resp = &LoggingResponseWriter{
+			ResponseWriter: ctx.Resp,
+		}
+		rw := ctx.Resp.(*LoggingResponseWriter)
```

Comment:
instead of directly assigning `&LoggingResponseWriter{...` to `ctx.Resp` and then having to type assert, wouldn't it be nicer to initialize that into a variable which then gets assigned into `ctx.Resp` like 

```
rw = LoggingResponseWriter{
			ResponseWriter: ctx.Resp,
		}
ctx.Resp = &rw
```

Not sure if that's just me, but if possible I prefer to avoid type assertions.

Reply:
Thanks! Fixed in 3e51444